componentize_cpy.py

```diff
@@ -0,0 +1,405 @@
+import argparse
+import base64
+import json
+import os
+import re
+import sys
+import tempfile
+import urllib.request, urllib.parse, urllib.error
+import zipfile
+from io import StringIO
+
+import clr
+import System
+import System.IO
+
+
+SCRIPT_COMPONENT_GUID = System.Guid("c9b2d725-6f87-4b07-af90-bd9aefef68eb")
+CPY_VER = "3.-1"
+TEMPLATE_VER = re.compile("{{version}}")
+TEMPLATE_NAME = re.compile("{{name}}")
+TEMPLATE_GHUSER_NAME = re.compile("{{ghuser_name}}")
+
+TYPES_MAP = dict(
+    none="6a184b65-baa3-42d1-a548-3915b401de53",
+    ghdoc="1c282eeb-dd16-439f-94e4-7d92b542fe8b",
+    float="9d51e32e-c038-4352-9554-f4137ca91b9a",
+    bool="d60527f5-b5af-4ef6-8970-5f96fe412559",
+    int="48d01794-d3d8-4aef-990e-127168822244",
+    complex="309690df-6229-4774-91bb-b1c9c0bfa54d",
+    str="3aceb454-6dbd-4c5b-9b6b-e71f8c1cdf88",
+    datetime="09bcf900-fe83-4efa-8d32-33d89f7a3e66",
+    guid="5325b8e1-51d7-4d36-837a-d98394626c35",
+    color="24b1d1a3-ab79-498c-9e44-c5b14607c4d3",
+    point="e1937b56-b1da-4c12-8bd8-e34ee81746ef",
+    vector="15a50725-e3d3-4075-9f7c-142ba5f40747",
+    plane="3897522d-58e9-4d60-b38c-978ddacfedd8",
+    interval="589748aa-e558-4dd9-976f-78e3ab91fc77",
+    uvinterval="74c906f3-db02-4cea-bd58-de375cb5ae73",
+    box="f29cb021-de79-4e63-9f04-fc8e0df5f8b6",
+    transform="c4b38e4c-21ff-415f-a0d1-406d282428dd",
+    line="f802a8cd-e699-4a94-97ea-83b5406271de",
+    circle="3c5409a1-3293-4181-a6fa-c24c37fc0c32",
+    arc="9c80ec18-b48c-41b0-bc6e-cd93d9c916aa",
+    polyline="66fa617b-e3e8-4480-9f1e-2c0688c1d21b",
+    rectangle="83da014b-a550-4bf5-89ff-16e54225bd5d",
+    curve="9ba89ec2-5315-435f-a621-b66c5fa2f301",
+    mesh="794a1f9d-21d5-4379-b987-9e8bbf433912",
+    surface="f4070a37-c822-410f-9057-100d2e22a22d",
+    subd="20f4ca9c-6c90-4fd6-ba8a-5bf9ca79db08",
+    brep="2ceb0405-fdfe-403d-a4d6-8786da45fb9d",
+    geometrybase="c37956f4-d39c-49c7-af71-1e87f8031b26"
+)
+
+EXPOSURE = dict(valid=set([-1, 2, 4, 8, 16, 32, 64, 128]), default=2)
+ACCESS = dict(valid=set([0, 1, 2]), map=dict(item=0, list=1, tree=2), default=0)
+PARAM_TYPE = dict(
+    valid=set(TYPES_MAP.values()), map=TYPES_MAP, default=TYPES_MAP["ghdoc"]
+)
+WIRE_DISPLAY = dict(
+    valid=set([0, 1, 2]), map=dict(default=0, faint=1, hidden=2), default=0
+)
+
+
+def fetch_ghio_lib(target_folder="temp"):
+    """Fetch the GH_IO.dll library from the NuGet packaging system."""
+    ghio_dll = "GH_IO.dll"
+    filename = "lib/net48/" + ghio_dll
+
+    response = urllib.urlopen("https://www.nuget.org/api/v2/package/Grasshopper/")
+    dst_file = os.path.join(target_folder, ghio_dll)
+    zip_file = zipfile.ZipFile(StringIO.StringIO(response.read()))
```

Comment:
```suggestion
    zip_file = zipfile.ZipFile(BytesIO(response.read()))
```

Reply:
Alright, I am leaving the `metadata.json` for the ipy examples with capital though.